service/controller/resource/ipam/azure_machinepool_checker.go

```diff
@@ -0,0 +1,80 @@
+package ipam
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	capzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	capiv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type AzureMachinePoolCheckerConfig struct {
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+}
+
+type AzureMachinePoolChecker struct {
+	ctrlClient client.Client
+	logger     micrologger.Logger
+}
+
+func NewAzureMachinePoolChecker(config AzureMachinePoolCheckerConfig) (*AzureMachinePoolChecker, error) {
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	a := &AzureMachinePoolChecker{
+		ctrlClient: config.CtrlClient,
+		logger:     config.Logger,
+	}
+
+	return a, nil
+}
+
+func (c *AzureMachinePoolChecker) Check(ctx context.Context, namespace string, name string) (bool, error) {
+	azureMachinePool := &v1alpha3.AzureMachinePool{}
+	err := c.ctrlClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, azureMachinePool)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	cluster, err := util.GetClusterFromMetadata(ctx, c.ctrlClient, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	azureCluster, err := c.getAzureClusterFromCluster(ctx, cluster)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	// In case there is no subnet tracked so far, we want to proceed with the allocation process.
+	for _, subnet := range azureCluster.Spec.NetworkSpec.Subnets {
+		if subnet.Name == azureMachinePool.Name {
+			return false, nil
```

Comment:
I would expect `Check` method to return `true` when the subnet is allocated, this seems a bit reversed. WDYT @tuommaki?

Reply:
Hmm now when I look at `Checker` interface, it says:
> Checker determines whether a subnet has to be allocated.

Reply:
I would like us to adjust the naming a bit, so that the purpose of interfaces, structs and functions is a bit more clear.

Reply:
This is currently following the structure of `aws-operator`. If we change the structure of this, I'd suggest considering the problem from the beginning. For me this feels a bit complicated considering what it actually does. Important thing here (in IPAM) is that necessary operations are executed within acquired lock.